store/src/java/com/zimbra/cs/lmtpserver/ExternalEmailWarning.java

```diff
@@ -172,27 +184,130 @@ public boolean isExternal(String receiverAddress, String originatorAddress) {
+                    }
+                }
+            } else if (ctTextPlainMatcher.find()) {
                 final int end = ctTextPlainMatcher.end();
-                final StringBuilder sb = new StringBuilder();
-                sb.append(content.substring(0, end));
-                if (getTextPlainWarning() != null)
-                    sb.append(getTextPlainWarning());
-                sb.append(content.substring(end));
-                content = sb.toString();
+                content = appendPlainTextWarning(content, end);
             }
             final Matcher ctTextHtmlMatcher = CONTENT_TYPE_TEXT_HTML_PATTERN.matcher(content);
-            if (ctTextHtmlMatcher.find()) {
-                final int end = ctTextHtmlMatcher.end();
-                final StringBuilder sb = new StringBuilder();
-                sb.append(content.substring(0, end));
-                if (getTextHtmlWarning() != null)
-                    sb.append(getTextHtmlWarning());
-                sb.append(content.substring(end));
-                content = sb.toString();
+            final Matcher ctTextHtmlBase64EncodedMatcher = CONTENT_TYPE_TEXT_HTML_BASE64_ENCODED_PATTERN.matcher(content);
+            if (ctTextHtmlBase64EncodedMatcher.find()) {
+                String mimeBoundary = getMimeBoundary(content);
+                if (mimeBoundary != null) {
+                    final int htmlBodyStart = ctTextHtmlBase64EncodedMatcher.end();
+                    String decodedHtmlContentMime = decodeMimeSection(content, htmlBodyStart, mimeBoundary);
+                    decodedHtmlContentMime = getUpdatedHtmlSection(decodedHtmlContentMime, htmlBodyStart);
+                    content = encodeMimeSection(decodedHtmlContentMime, htmlBodyStart, mimeBoundary);
+                }
+            } else if (ctTextHtmlMatcher.find()) {
+                final int htmlBodyStart = ctTextHtmlMatcher.end();
+                content = getUpdatedHtmlSection(content, htmlBodyStart);
             }
         }
         return content;
     }
+    /**
+     * Updates the text/html section of MIME
+     * Checks weather html section contains <body> tag and appends the warning, else appends the warning at start of section
+     * */
+    public String getUpdatedHtmlSection(String content, int htmlBodyStartIndex) {
+        Matcher containsBody = CONTAINS_BODY_TAG_PATTERN.matcher(content.substring(htmlBodyStartIndex));
+        if (containsBody.find()) {
+            final int end = containsBody.end() + htmlBodyStartIndex;
+            content = appendHtmlWarning(content, end);
+        } else {
+            final int end = htmlBodyStartIndex;
+            content = appendHtmlWarning(content, end);
+        }
+        return content;
+    }
+
+    /**
+     * Appends HTML warning at a specific index and returns the MIME
+     * */
+    public String appendHtmlWarning(String content, int index) {
+        final StringBuilder sb = new StringBuilder();
```

Comment:
appendHtmlWarning and appendPlainTextWarning can be combined in one method (Lets say appendWarning()). add additional parameter whether to check for plain or html text.

Reply:
Refactored the code and created addWarning method to add warning of any type